cpp/scenario/gazebo/include/scenario/gazebo/Joint.h

```diff
@@ -0,0 +1,185 @@
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SCENARIO_GAZEBO_JOINT_H
+#define SCENARIO_GAZEBO_JOINT_H
+
+#include <ignition/gazebo/Entity.hh>
+#include <ignition/gazebo/EntityComponentManager.hh>
+#include <ignition/gazebo/EventManager.hh>
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace scenario {
+    namespace base {
+        struct PID;
+        struct Limit;
+        enum class JointType
+        {
+            Invalid,
+            Fixed,
+            Revolute,
+            Prismatic,
+            Ball,
+        };
+        enum class JointControlMode
+        {
+            Idle,
+            Force,
+            Velocity,
+            Position,
+            PositionInterpolated,
+        };
+    } // namespace base
+    namespace gazebo {
+        class Joint;
+        class Model;
+    } // namespace gazebo
+} // namespace scenario
+
+class scenario::gazebo::Joint
+{
+public:
+    Joint();
+    virtual ~Joint();
+
+    // ============
+    // Gazebo Joint
+    // ============
+
+    uint64_t id() const;
+    bool initialize(const ignition::gazebo::Entity jointEntity,
+                    ignition::gazebo::EntityComponentManager* ecm,
+                    ignition::gazebo::EventManager* eventManager);
+    bool createECMResources();
+
+    bool historyOfAppliedJointForcesEnabled() const;
+    bool enableHistoryOfAppliedJointForces( //
+        const bool enable,
+        const size_t maxHistorySize = 100);
+    std::vector<double> getHistoryOfAppliedJointForces() const;
+
+    // ==========
+    // Joint Core
+    // ==========
+
+    size_t dofs() const;
+    std::string name() const;
+    base::JointType type() const;
+
+    base::Limit positionLimit() const;
+
+    base::JointControlMode controlMode() const;
+    bool setControlMode(const base::JointControlMode mode);
+
+    double controllerPeriod() const;
+
+    double maxEffort() const;
+    bool setMaxEffort(const double maxEffort);
+
+    base::PID pid() const;
+    bool setPID(const base::PID& pid);
+
+    // ==================
+    // Single DOF methods
+    // ==================
+
+    double force(const size_t dof = 0) const;
```

Comment:
I  suggest to document this method.  First of all, I understand the idea of  using "force" to be joint type-independent, but this will be clearly confusing for most roboticist used to the concept of "joint torques". Documenting this will avoid confusion. An  additional common source of confusion is the difference the torque that the simulation engine takes in input, from which the **friction torques** are removed to obtain the actual torque delivered to the multibody structure, and the torque tipically used in whole-body control, that instead just refer to the actual torque delivered to the multibody structure.

In formulas, for a fixed manipulator in simulation the "torque" is typically referred to:
$$
\tau_{sim} = M(q)\ddot{q} +  C(q,\dot{q})\dot{q} + J^T f_{ext} - \tau_{friction}(q, \dot{q}) 
$$
while in whole-body controllers with "torque" we typically refer to: 
$$
\tau_{control} = M(q)\ddot{q} +  C(q,\dot{q})\dot{q} + J^T f_{ext} 
$$
I think we should be clear on which force we are referring to, see https://bitbucket.org/osrf/gazebo/issues/1321/proper-joint-torque-feedback for a related issue.

cc @DanielePucci

Reply:
Thanks a lot for this comment! The past is coming back as far as I see :laughing: That issue in enlightening. So far, I never considered friction in the design of the APIs. Now I am realizing that I should have.

> An additional common source of confusion is the difference the torque that the simulation engine takes in input, from which the friction torques are removed to obtain the actual torque delivered to the multibody structure, and the torque tipically used in whole-body control, that instead just refer to the actual torque delivered to the multibody structure.

Right now `Joint::force()` is returning the force that gazebo has sent to the physics engine in the previous step (i.e. the `JointForceCmd` that could be filled by PID or custom controllers). The logic was introduced in #141. After the discussion of https://github.com/robotology/gazebo-yarp-plugins/issues/471 we are aligned to think that it could be confusing.

I just checked and in the `Physics::PostUpdate` method we could [ask the joint force using `physics::Joint::getForce`](https://bitbucket.org/ignitionrobotics/ign-physics/src/b5540de53faf792d5cda2087b3b7c91008e38a82/include/ignition/physics/Joint.hh#lines-62:68) to the physics engine. Unfortunately it will return the value previously set with `physics::Joint::setForce`, so the value will also contain the fraction lost in friction. Due to this lack, I think it's better to remove `Joint::force` and `Joint::setForce`.

I propose the following APIs:

```cpp
// Get the motor force references (content of existing `JointForceCmd` component).
double forceTarget(const size_t dof = 0) const;
std::vector<double> jointForceTarget() const;

// Set the joint force references (filling the `JointForceCmd` component).
// Controllers will use this method to set their output.
// TODO: should this force be additive?
bool setForceTarget(const double force, const size_t dof = 0);
bool setJointForceTarget(const std::vector<double>& force);

// Joint torques should be implemented through FT sensor plugins
// that are in upstream's TODO list. If we want to avoid having too many
// threads (one per joint) we could also consider to move the logic in this
// Joint class or in the Physics system.
// bool force() const
// std::vector<double> jointForce() const;

// Return the reduction ratio (default = 1.0) that can be filled by a motor plugin
double gearReductionRatio() const;

// Setting the max force should happen in the motor side. Since users can get the
// gear ratio, I prefer having the following rather than having e.g. `maxForce()`.
// PIDs and custom controllers, that are operating in the joint side, will use  
// these limits (scaled by the gear ratio) to clip their output.
double maxMotorForce(const size_t dof = 0) const;
std::vector<double> jointMaxMotorForce() const;
bool setMaxMotorForce(const double maxForce);
bool setJointMaxMotorForce(const std::vector<double>& maxForce);

// TODO: not implemented now. If we need them in the future we can extend
//       the Joint APIs.
// double motorForceTarget(const size_t dof = 0) const;
// std::vector<double> jointMotorForceTarget() const;
// bool setMotorForceTarget(const double force, const size_t dof = 0);
// bool setJointMotorForceTarget(const std::vector<double>& force);

// These methods are meant to simplify getting the values to be used to
// calculate the usual regularization term that minimizes joint torques.
// They will return the content of `JointForceCmd` over a window.
bool historyOfAppliedJointForcesEnabled() const;
bool enableHistoryOfAppliedJointForces(const bool enable,
                                       const size_t maxHistorySize = 100);
std::vector<double> getHistoryOfAppliedJointForces() const;
```

What do you think @traversaro?

I tried to take into account that multiple Systems could operate on a joint in a single simulation step. The most complex scenario is the following:

- The joint is controlled in `JointControlMode::Force` using a custom controller (let's assume it is a whole-body controller implemented as a model plugin System) and has a plugin that simulates the second order dynamics (it could be either model-based with motor inertia, gearbox ratio and delay, or an actuatornet).
  1. The user sets joint references through the Model APIs (position, velocity, and acceleration). The `Joint{Position,Velocity,Acceleration}Target` components are filled.
  1. In the `PreUpdate` step,  the custom controller System processes the targets and calls `Joint::setJointForce`, that fills the `JointForceCmd` component.
  1. In the `PreUpdate` step [1] the second order dynamics System reads the `JointForceCmd` component and changes it, e.g. reducing it due to the motor inertia and delaying it.
  1. In the `Update` step the physics System reads the `JointForceCmd` and sets this value in the physics engine.

[1] Assuming we can fix the order of systems execution, that now is related to the loading order. I have to double check how this is processed.

Note: I just realized something very cool. Considering that: 1) we can decide the rate of Systems by setting the simulator update rate, 2) we control the physics system since it's vendored, if 3) we allow the physics system to run slower than the simulator (skipping few iterations), **we can have the system that processes the second order dynamics running faster than the physics**. @DanielePucci I remember that this was one of the limitation that we had when we tried to setup the previous co-simulation architecture for motor's second-order dynamics.

Reply:
> If we want to avoid having too many threads (one per joint)

Do you have any docs/code/issue pointer on the fact that a thread is created for each system? This is a critical performance bottleneck. 

> // Return the reduction ratio (default = 1.0) that can be filled by a motor plugin
> double gearReductionRatio() const;

I am a bit confused by this. Is this something that exists in Ignition Gazebo? If instead is a parameter of a possible future motor/transmission model, then I have two observations: 
* The motor plugins will have a lot of parameters, why expose only one in the main API? 
* How can a parameter in the motor plugin influence of the value returned by Ignition Gazebo-related  API? 


> we can have the system that processes the second order dynamics running faster than the physics. @DanielePucci I remember that this was one of the limitations we had when we tried to setup the previous co-simulation architecture for motor's second-order dynamics.

Not sure what you mean here. If you run co-simulation using the Gazebo physics as "master", you can easily run a second order "faster" (I guess this means with "shorter" timestep) by integrating it multiple times during the physics engine update timestep. For example, if hte main physics has a timestep of 1 ms, and you want to integrate the second order with a fixed integrator with a timestep of 0.1 ms, you just need to step the integrator of the second order dynamics 10 times in the callback of the main physics thread.

Reply:
> we control the physics system since it's vendored

I totally understand the convenience of using the vendored physics system, but I think for second order  dynamics ideally/if possible we should use a solution that works also with the upstream physics system, so that we can then re-use easily with systems simulated always with Ignition Gazebo but that interface with controllers in other ways, for example using gazebo-yarp-plugins.

Reply:
> // Joint torques should be implemented through sensor plugins that
// provide a measurement in the joint side. If we want to avoid having too many
// threads (one per joint) we could also consider to move the logic in this
// Joint class or in the Physics system.
// bool force() const
// std::vector<double> jointForce() const;

Why joint torques should be implemented via sensor plugins? I suspect that the solution suggested in https://bitbucket.org/osrf/gazebo/issues/1321/proper-joint-torque-feedback (i.e. get the wrench exchanged between two bodies and project it on the joint axis) should work without the need of custom plugins.

Reply:
> Do you have any docs/code/issue pointer on the fact that a thread is created for each system? This is a critical performance bottleneck.

Given the architecture of Ignition Gazebo, it seems to me that everything that implements a System (sensors, model and world plugins, ...) run in a different thread. I am not yet familiar with sensors, I can have a look to their implementation.

>> // Return the reduction ratio (default = 1.0) that can be filled by a motor plugin
double gearReductionRatio() const;
> I am a bit confused by this. Is this something that exists in Ignition Gazebo?

No it does not.

> If instead is a parameter of a possible future motor/transmission model, then I have two observations:
> - The motor plugins will have a lot of parameters, why expose only one in the main API?
> - How can a parameter in the motor plugin influence of the value returned by Ignition Gazebo-related API?

The main reason of the combination of `gearReductionRatio` and the `*Motor*` methods is to have less confusing APIs. If you read just the name of the method, you instantly get its semantics. Do you think that replacing them with `Joint::maxJointForce` and `Joint::setMaxJointForce` would be better?

> Not sure what you mean here. If you run co-simulation using the Gazebo physics as "master", you can easily run a second order "faster"

Yep, I agree with you. It's the same.

> I totally understand the convenience of using the vendored physics system, but I think for second order dynamics ideally/if possible we should use a solution that works also with the upstream physics system, so that we can then re-use easily with systems simulated always with Ignition Gazebo but that interface with controllers in other ways, for example using gazebo-yarp-plugins.

Yes maybe I did not explain it enough clearly in my comment but in the bullet point of the example I wrote that the second-order dynamics is indeed a standalone plugin that provides a System. The comment on the vendored physics was only for the stepping strategy that, as quoted above, could be avoided.

Reply:
> The main reason of the combination of gearReductionRatio and the *Motor* methods is to have less confusing APIs. If you read just the name of the method, you instantly get its semantics. Do you think that replacing them with Joint::maxJointForce and Joint::setMaxJointForce would be better?

Sorry, it is not really clear to me how the gearReductionRatio  and the `maxMotorForce`  would be interacting. Just to clarify, what I have personally in mind is the following https://app.diagrams.net?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&title=SecondOrderDynamicsIntegration.drawio#R7VpZk5s4EP41rso%2B2MVhbPw4PibZ2mR3Nt7KTp5SMsi2dgERIcZ2fn1aIAwIfIyvkMn4xajRRffXX7caWubIX79lKFx%2BoC72WobmrlvmuGUYumba8Cckm1Ri9a1UsGDElZ1ywZR8w9lIKY2Ji6NSR06px0lYFjo0CLDDSzLEGF2Vu82pV141RAtcEUwd5FWl%2FxKXL1OpbWm5%2FB0mi2W2sq7JOz7KOktBtEQuXRVE5qRljhilPL3y1yPsCeVleknH3e%2B4u90YwwE%2FZsDnvwfzL3%2FGjw%2FzjxNr%2BIe1mX0ateUsT8iL5QNPMajSBdlfzMUM%2FsebAPnEkQ%2FBN5lmVkvC8TREjmivwPotc7jkvgctHS5RFKb2mJM1hi0Mq%2FvNFseM43VBJPf%2FFlMfc7aBLvKuaUpdSjAZmbJXuWn0TN%2FLolkyIZJwWGznzjUGF1Jpz1CgUVFgRVE4cO8EEqHleCiKQJclTe1UDHZL2KyqpfDYVs1TZzKGPcTJUxnRdZqQKzxQAjvZar1rlbVuaooyIxozB8tRRQQqE1nmgYk4YgvMKxMlhtk%2B9um20uuM1fNAW0OXPMHlQlySIIz5PYUnym7CYoX7NUNgzjekgzstAbPe11i49NCnnLKErtjXGOfyXbMqqAGX4IpHeWQRCBABTsAzzaFwHAJUdSdv%2BMR1xfAhwxH5hmbJVBq0Q6HRRMfWsGWNQeKhGfaGyPl%2FwWgcuCPqwV7FuuY8%2BUGXOQ245GO7FqV7%2FUH16S1Dy22VSLDO19tax8xihcRLxlZn4lkvzzooj6fzeYSvAj%2Bzgr4PMQSyIXXFYz8sN8AMEVxNggUJcBP4VnXYBvCt9fL5VrcVrfdO5NtKuLQGHeumjNurQj5lxft%2FGAoin4B1aAA9fhekm9JnQo4zVrJpxp2Ck9pRQkp30MEO1ymh5tSakOkbh8aeyGFmWKwO6aLI8oyRYKGYJXZPGhk1jy62sotDgB8JFiBAvnDRYBaFaYhIHpQvxZZ8kSf%2F9lI4P%2FPJS3C%2BpVC%2BeaEUpjOwB4VfeRW939GVOa8XBuxjkpAcWe%2Fpqu3hJ3GsOpRBVGYZgTkZ9TyRxx%2BVdPyAGNNrXIgZnB9iFD%2FCa8IfpV%2BK68%2FiGqg4bY3XhVvjTdYA020e5XxJozBKNPNhSSsbt9MkadQ4nJ%2BkIWFfGt2oaNnVlEOFiotjo6V6zDH6Nz6daC8Tdw3BSSWX7V4KJ%2BpE18ZJtWbzipMr4uRifKJOdG2cVM%2Bb%2F4m17hwHQ0YAOhIJ6U%2BXge5KMJ6dbHYPHbOaW0nQu%2FWm%2FYQ96hC%2BeTXrHnpusFmrxY3ErA80Iq%2FeeoBNG2zWahXk7IANsXIqR5Mo2SFxJrkwj85ax7BKAfrU8HzwWGA3K4wbF0r3dGUiXT8ujIM50abQTfrXzg2bXWWd8us8uEhnvCw064oRt4Nm%2F5eE5kEqOxqaytHXuBY0lQ3rvVtA8wplmBx5JeDpvwjwVCrrXuwIfONSSfYcBWx8xHMM9nFwVEFJk7IkOJQdXVbfesAl6up6r9e%2FDHuVTd9Wyet6eZRRV%2FhIy9pCw60jXtvoVriuvrR5B56WfOryPq2za1Meg%2BHyknk6%2FY6aOeychJEAS7REIU64iMaiOn6gmn5y8rz7%2BxjFv7N4UKAXs7aWvu%2BMdN7nMcY1Wdy2W6Xak3Go%2BLQtdG2LW1v%2B31%2FoesCMgEYEWdyoyt6wspiuYss6NWlRPwIybxw7qmWxmjdnUxxEydc79xi7M2D3n%2FCrnefEmq2fNirWWFbH7pXQovc7g37xRW7%2F%2FOgDzfwrzLR7%2Fi2rOfkO  .

Reply:
> Sorry, it is not really clear to me how the gearReductionRatio and the maxMotorForce would be interacting.

I had in mind:

$$
\tau^{Max}_{joint} = T * \tau^{Max}_{motor}
$$

Reply:
Hi guys, I am not understanding why the terminology  "force" should be independent from the joint-type, since if you tell me "force", I associate Newtons to it, so prismatic joint not revolute.  Probably you can help me with that

Concerning second order dynamics, as I was mentioning today, I would like to converge towards a model exchange based simulation infrastructure 

CC
@diegoferigo @paolo-viceconte @CarlottaSartore

Reply:
> 
> 
> > Sorry, it is not really clear to me how the gearReductionRatio and the maxMotorForce would be interacting.
> 
> I had in mind:
> 
> $$
\tau_{joint}^{Max} = T * \tau_{motor}^{Max}
$$

But in that case the Ignition Gazebo API will return for "\tau_{motor}" a completly different  quantity then  Gym Ignition. I think that this kind of "hidden"  logic  do not  aid in clarity.

Reply:
> Hi guys, I am not understanding why the terminology "force" should be independent from the joint-type, since if you tell me "force", I associate Newtons to it, so prismatic joint not revolute. Probably you can help me with that

In this domain we're talking of _generalized forces_. We can add the extra `Generalized` prefix to those methods if this makes them more clear, with the cost of APIs more verbose. Another terminology that could be used is `Effort` instead of `Force`, but it seems to me to be way less common.

> But in that case the Ignition Gazebo API will return for "\tau_{motor}" a completly different quantity then Gym Ignition. I think that this kind of "hidden" logic do not aid in clarity.

Ok let's ignore this gear ratio than. My main point was that typically you know the maximum torque that a it could exerted at the motor level, not the joint level. Removing the `Motor` methods, if users have a motor plugin and want to set its maximum torque, they need to can only set the joint maximum torque calculated from the user code using the gear ratio (gathered somehow from the custom motor plugin, or hardcoded).

Here below the updated proposal:

```cpp
// Get the motor force references (content of existing `JointForceCmd` component).
double generalizedForceTarget(const size_t dof = 0) const;
std::vector<double> jointGeneralizedForceTarget() const;

// Set the joint force references (filling the `JointForceCmd` component).
// Controllers will use this method to set their output.
// TODO: should this force be additive?
bool setGeneralizedForceTarget(const double force, const size_t dof = 0);
bool setJointGeneralizedForceTarget(const std::vector<double>& force);

// Joint torques should be implemented through sensor plugins that
// provide a measurement in the joint side. If we want to avoid having too many
// threads (one per joint) we could also consider to move the logic in this
// Joint class or in the Physics system.
// bool generalizedForce() const
// std::vector<double> jointGeneralizedForce() const;

double maxGeneralizedForce(const size_t dof = 0) const;
std::vector<double> jointMaxGeneralizedForce() const;
bool setMaxGeneralizedForce(const double maxForce);
bool setJointMaxGeneralizedForce(const std::vector<double>& maxForce);

// These methods are meant to simplify getting the values to be used to
// calculate the usual regularization term that minimizes joint torques.
// They will return the content of `JointForceCmd` over a window.
bool historyOfAppliedJointGeneralizedForcesEnabled() const;
bool enableHistoryOfAppliedJointGeneralizedForces(
    const bool enable, const size_t maxHistorySize = 100);
std::vector<double> getHistoryOfAppliedJointGeneralizedForces() const;
```

Reply:
Updated APIs in 7709620d081da159b599ebbec0278ec816bcd8d3